client/local/specio/resource_spec_io.go

```diff
@@ -0,0 +1,76 @@
+package specio
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+
+	"github.com/odpf/optimus/client/local"
+	"github.com/odpf/optimus/client/local/internal"
+	"github.com/odpf/optimus/client/local/model"
+)
+
+type resourceSpecReadWriter struct {
+	referenceSpecFileName string
+	specFS                afero.Fs
+}
+
+func NewResourceSpecReadWriter(specFS afero.Fs) (local.SpecReadWriter[*model.ResourceSpec], error) {
+	if specFS == nil {
+		return nil, errors.New("spec fs is nil")
+	}
+	return &resourceSpecReadWriter{
+		referenceSpecFileName: "resource.yaml",
+		specFS:                specFS,
+	}, nil
+}
+
+func (r resourceSpecReadWriter) ReadAll(rootDirPath string) ([]*model.ResourceSpec, error) {
+	if rootDirPath == "" {
+		return nil, errors.New("root dir path is empty")
+	}
+
+	specDirPaths, err := internal.DiscoverSpecDirPaths(r.specFS, rootDirPath, r.referenceSpecFileName)
+	if err != nil {
+		return nil, fmt.Errorf("error discovering spec paths under [%s]: %w", rootDirPath, err)
+	}
+
+	specs := make([]*model.ResourceSpec, len(specDirPaths))
+	for i, dirPath := range specDirPaths {
+		filePath := filepath.Join(dirPath, r.referenceSpecFileName)
+		spec, err := internal.ReadSpec[*model.ResourceSpec](r.specFS, filePath)
+		if err != nil {
+			return nil, fmt.Errorf("error reading spec under [%s]: %w", filePath, err)
+		}
+		specs[i] = spec
+	}
+	return specs, nil
+}
+
+func (r resourceSpecReadWriter) ReadByName(rootDirPath, name string) (*model.ResourceSpec, error) {
+	if name == "" {
+		return nil, errors.New("name is empty")
+	}
+	allSpecs, err := r.ReadAll(rootDirPath)
```

Comment:
Are we reading all the specs here to read only a single spec. Is there any other reliable way to find the specs ?

Reply:
agreed that reading all may be not the optimal solution we are looking for. but, the tricky part is, we are searching by the name. and no other place we can get the name information other than reading the whole file. the closest we can do is we can limit it. for example, we read the specs one by one and stop after finding what we are looking for. it can reduce the number of file we have to read. what do you think?